File: leetcode/384-shuffle-an-array.py
# https://leetcode.com/problems/shuffle-an-array/ (top-interview-questions)

import random
from typing import List

class Solution:
    def __init__(self, nums: List[int]):
        self.original = nums

    def reset(self) -> List[int]:
        return self.original

    def shuffle(self) -> List[int]:
        shuffled = self.original.copy()
        for i in range(len(shuffled)):
            random_idx = random.randint(i, len(shuffled)-1)
            shuffled[i], shuffled[random_idx] = shuffled[random_idx], shuffled[i]
        return shuffled
        #------------cheating------------
        # shuffled = self.original.copy()
        # random.shuffle(shuffled)
        # return shuffled

File: leetcode/384-shuffle-an-array.js
// https://leetcode.com/problems/shuffle-an-array/ (top-interview-questions)

/**
 * @param {number[]} nums
 */
var Solution = function(nums) {
  this.original = nums;
  this.shuffled = [...nums];
};

/**
 * @return {number[]}
 */
Solution.prototype.reset = function() {
  return this.original;
};

/**
 * @return {number[]}
 */
Solution.prototype.shuffle = function() {
  this.shuffled.sort( _ => Math.random() - 0.5);
  return this.shuffled;
};

/** 
 * Your Solution object will be instantiated and called as such:
 * var obj = new Solution(nums)
 * var param_1 = obj.reset()
 * var param_2 = obj.shuffle()
 */
